File: examples/python/agent_anthropic.py
#!/usr/bin/env python3
"""
NCHE Agent Example - Anthropic Claude

This example demonstrates how to build an AI agent using Anthropic's Claude
that executes actions through NCHE for human oversight.

Requirements:
    pip install anthropic requests

Usage:
    export ANTHROPIC_API_KEY=your_api_key
    export NCHE_API_KEY=your_nche_api_key
    python agent_anthropic.py
"""

import json
import os
from anthropic import Anthropic
from nche_client import NcheClient

# Configuration
NCHE_URL = os.getenv("NCHE_URL", "http://localhost:3000")
NCHE_API_KEY = os.getenv("NCHE_API_KEY")
ANTHROPIC_API_KEY = os.getenv("ANTHROPIC_API_KEY")

# Define tools that the agent can use (these map to NCHE tools)
TOOLS = [
    {
        "name": "send_email",
        "description": "Send an email to a recipient. Use this when you need to communicate via email.",
        "input_schema": {
            "type": "object",
            "properties": {
                "to": {
                    "type": "string",
                    "description": "Email address of the recipient",
                },
                "subject": {
                    "type": "string",
                    "description": "Subject line of the email",
                },
                "body": {
                    "type": "string",
                    "description": "Body content of the email",
                },
            },
            "required": ["to", "subject", "body"],
        },
    },
    {
        "name": "http_request",
        "description": "Make an HTTP request to an API endpoint.",
        "input_schema": {
            "type": "object",
            "properties": {
                "method": {
                    "type": "string",
                    "enum": ["GET", "POST", "PUT", "DELETE"],
                    "description": "HTTP method",
                },
                "url": {
                    "type": "string",
                    "description": "URL to request",
                },
                "headers": {
                    "type": "object",
                    "description": "HTTP headers",
                },
                "body": {
                    "type": "string",
                    "description": "Request body (for POST/PUT)",
                },
            },
            "required": ["method", "url"],
        },
    },
]


def execute_tool_via_nche(nche: NcheClient, tool_name: str, tool_input: dict) -> str:
    """Execute a tool through NCHE and return the result."""
    print(f"\n[NCHE] Proposing action: {tool_name}")
    print(f"[NCHE] Parameters: {json.dumps(tool_input, indent=2)}")

    action = nche.execute_tool(tool_name, tool_input, wait=True, timeout=120)

    print(f"[NCHE] Action state: {action.state}")

    if action.state == "executed":
        return json.dumps(action.result or {"status": "success"})
    elif action.state == "denied":
        return json.dumps({"error": "Action was denied by human reviewer"})
    elif action.state == "failed":
        return json.dumps({"error": action.error or "Action execution failed"})
    else:
        return json.dumps({"error": f"Unexpected state: {action.state}"})


def run_agent(user_message: str):
    """Run the agent with a user message."""
    if not NCHE_API_KEY:
        raise ValueError("NCHE_API_KEY environment variable is required")
    if not ANTHROPIC_API_KEY:
        raise ValueError("ANTHROPIC_API_KEY environment variable is required")

    # Initialize clients
    nche = NcheClient(NCHE_URL, NCHE_API_KEY)
    anthropic = Anthropic(api_key=ANTHROPIC_API_KEY)

    # Create NCHE session
    print("[NCHE] Creating session...")
    session = nche.create_session(
        actor_id="claude-agent",
        actor_type="agent",
        autonomy_level="supervised",  # Require human approval
    )
    print(f"[NCHE] Session created: {session.id}")

    try:
        # Initialize conversation
        messages = [{"role": "user", "content": user_message}]

        print(f"\n[User] {user_message}\n")

        # Agent loop
        while True:
            # Call Claude
            response = anthropic.messages.create(
                model="claude-sonnet-4-20250514",
                max_tokens=4096,
                system="""You are a helpful assistant that can send emails and make HTTP requests.
When you need to perform an action, use the provided tools.
All actions are reviewed by humans before execution for safety.""",
                tools=TOOLS,
                messages=messages,
            )

            # Check if we're done
            if response.stop_reason == "end_turn":
                # Extract final text response
                for block in response.content:
                    if hasattr(block, "text"):
                        print(f"[Claude] {block.text}")
                break

            # Process tool uses
            tool_uses = []
            text_content = []

            for block in response.content:
                if block.type == "tool_use":
                    tool_uses.append(block)
                elif hasattr(block, "text"):
                    text_content.append(block.text)

            # Print any text
            for text in text_content:
                print(f"[Claude] {text}")

            # Execute tools through NCHE
            if tool_uses:
                # Add assistant message with tool uses
                messages.append({"role": "assistant", "content": response.content})

                tool_results = []
                for tool_use in tool_uses:
                    result = execute_tool_via_nche(
                        nche, tool_use.name, tool_use.input
                    )
                    tool_results.append(
                        {
                            "type": "tool_result",
                            "tool_use_id": tool_use.id,
                            "content": result,
                        }
                    )

                # Add tool results
                messages.append({"role": "user", "content": tool_results})
            else:
                # No more tool calls, we're done
                break

    finally:
        # Clean up session
        print("\n[NCHE] Ending session...")
        nche.end_session()
        print("[NCHE] Session ended")


if __name__ == "__main__":
    # Example: Ask the agent to send an email
    run_agent(
        "Please send an email to team@example.com with subject 'Weekly Update' "
        "and body 'Hi team, here is this week's progress update. All tasks are on track.'"
    )

File: examples/rust/src/agent_anthropic.rs
//! NCHE Agent Example - Anthropic Claude
//!
//! This example demonstrates how to build an AI agent using Anthropic's Claude
//! that executes actions through NCHE for human oversight.
//!
//! # Usage
//! ```bash
//! export ANTHROPIC_API_KEY=your_api_key
//! export NCHE_API_KEY=your_nche_api_key
//! cargo run --bin agent_anthropic
//! ```

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::env;
use std::time::Duration;

mod nche_client;
use nche_client::NcheClient;

// Anthropic API types
#[derive(Debug, Serialize)]
struct AnthropicRequest {
    model: String,
    max_tokens: u32,
    system: String,
    tools: Vec<Tool>,
    messages: Vec<Message>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
struct Tool {
    name: String,
    description: String,
    input_schema: Value,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
struct Message {
    role: String,
    content: MessageContent,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(untagged)]
enum MessageContent {
    Text(String),
    Blocks(Vec<ContentBlock>),
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "type")]
enum ContentBlock {
    #[serde(rename = "text")]
    Text { text: String },
    #[serde(rename = "tool_use")]
    ToolUse { id: String, name: String, input: Value },
    #[serde(rename = "tool_result")]
    ToolResult { tool_use_id: String, content: String },
}

#[derive(Debug, Deserialize)]
struct AnthropicResponse {
    content: Vec<ContentBlock>,
    stop_reason: String,
}

fn get_tools() -> Vec<Tool> {
    vec![
        Tool {
            name: "send_email".to_string(),
            description: "Send an email to a recipient.".to_string(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "to": { "type": "string", "description": "Email address" },
                    "subject": { "type": "string", "description": "Subject line" },
                    "body": { "type": "string", "description": "Email body" }
                },
                "required": ["to", "subject", "body"]
            }),
        },
        Tool {
            name: "http_request".to_string(),
            description: "Make an HTTP request.".to_string(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "method": { "type": "string", "enum": ["GET", "POST", "PUT", "DELETE"] },
                    "url": { "type": "string", "description": "URL to request" },
                    "headers": { "type": "object" },
                    "body": { "type": "string" }
                },
                "required": ["method", "url"]
            }),
        },
    ]
}

async fn call_anthropic(
    client: &reqwest::Client,
    api_key: &str,
    messages: &[Message],
) -> Result<AnthropicResponse> {
    let request = AnthropicRequest {
        model: "claude-sonnet-4-20250514".to_string(),
        max_tokens: 4096,
        system: "You are a helpful assistant that can send emails and make HTTP requests. \
                 All actions are reviewed by humans before execution.".to_string(),
        tools: get_tools(),
        messages: messages.to_vec(),
    };

    let response = client
        .post("https://api.anthropic.com/v1/messages")
        .header("x-api-key", api_key)
        .header("anthropic-version", "2023-06-01")
        .header("content-type", "application/json")
        .json(&request)
        .send()
        .await
        .context("Failed to call Anthropic API")?;

    if !response.status().is_success() {
        let error = response.text().await?;
        anyhow::bail!("Anthropic API error: {}", error);
    }

    response.json().await.context("Failed to parse response")
}

async fn execute_tool_via_nche(
    nche: &NcheClient,
    session_id: &str,
    tool_name: &str,
    tool_input: Value,
) -> Result<String> {
    println!("\n[NCHE] Proposing action: {}", tool_name);
    println!("[NCHE] Parameters: {}", serde_json::to_string_pretty(&tool_input)?);

    let action = nche.propose_action(session_id, tool_name, tool_input).await?;
    let action = nche.wait_for_action(&action.id, Duration::from_secs(120)).await?;

    println!("[NCHE] Action state: {}", action.state);

    Ok(match action.state.as_str() {
        "executed" => serde_json::to_string(&action.result.unwrap_or(json!({"status": "success"})))?,
        "denied" => json!({"error": "Action was denied by human reviewer"}).to_string(),
        "failed" => json!({"error": action.error.unwrap_or_else(|| "Action failed".to_string())}).to_string(),
        state => json!({"error": format!("Unexpected state: {}", state)}).to_string(),
    })
}

#[tokio::main]
async fn main() -> Result<()> {
    let nche_url = env::var("NCHE_URL").unwrap_or_else(|_| "http://localhost:3000".to_string());
    let nche_api_key = env::var("NCHE_API_KEY").context("NCHE_API_KEY is required")?;
    let anthropic_api_key = env::var("ANTHROPIC_API_KEY").context("ANTHROPIC_API_KEY is required")?;

    let http_client = reqwest::Client::new();
    let nche = NcheClient::new(&nche_url, &nche_api_key);

    // Create NCHE session
    println!("[NCHE] Creating session...");
    let session = nche.create_session("claude-agent", "agent", "supervised").await?;
    println!("[NCHE] Session created: {}", session.id);

    let user_message = "Please send an email to team@example.com with subject 'Weekly Update' \
                        and body 'Hi team, here is this week's progress update.'";

    println!("\n[User] {}\n", user_message);

    let mut messages = vec![Message {
        role: "user".to_string(),
        content: MessageContent::Text(user_message.to_string()),
    }];

    // Agent loop
    loop {
        let response = call_anthropic(&http_client, &anthropic_api_key, &messages).await?;

        // Check for text response
        for block in &response.content {
            if let ContentBlock::Text { text } = block {
                println!("[Claude] {}", text);
            }
        }

        // Check if done
        if response.stop_reason == "end_turn" {
            break;
        }

        // Extract tool uses
        let tool_uses: Vec<_> = response.content.iter()
            .filter_map(|b| {
                if let ContentBlock::ToolUse { id, name, input } = b {
                    Some((id.clone(), name.clone(), input.clone()))
                } else {
                    None
                }
            })
            .collect();

        if tool_uses.is_empty() {
            break;
        }

        // Add assistant message
        messages.push(Message {
            role: "assistant".to_string(),
            content: MessageContent::Blocks(response.content),
        });

        // Execute tools and collect results
        let mut tool_results = Vec::new();
        for (id, name, input) in tool_uses {
            let result = execute_tool_via_nche(&nche, &session.id, &name, input).await?;
            tool_results.push(ContentBlock::ToolResult {
                tool_use_id: id,
                content: result,
            });
        }

        // Add tool results
        messages.push(Message {
            role: "user".to_string(),
            content: MessageContent::Blocks(tool_results),
        });
    }

    // End session
    println!("\n[NCHE] Ending session...");
    nche.end_session(&session.id).await?;
    println!("[NCHE] Session ended");

    Ok(())
}
